include/votca/tools/NDimVector.h

```diff
@@ -0,0 +1,104 @@
+/*
+ *            Copyright 2009-2020 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef VOTCA_TOOLS_NDIMVECTOR_H
+#define VOTCA_TOOLS_NDIMVECTOR_H
+
+// Standard includes
+#include "types.h"
+#include <array>
+#include <cassert>
+#include <numeric>
+#include <vector>
+
+namespace votca {
+namespace tools {
+
+/**
+ * \brief N-Dim Vector
+ *
+ * This is basically a normal std::vector with an additional
+ * operator(i,j,k,l...), which allows you to treat it as a higher dim object
+ * It is column major (the first index is the fast one)
+ *
+ */
+
+template <class T, int dim>
+class NDimVector {
+
+  static constexpr int dim_ = dim;
+
+ public:
+  constexpr Index rank() { return dimensions_.size(); }
+
+  template <typename... IndexTypes>
+  NDimVector(Index firstDimension, IndexTypes... otherDimensions) {
+    // The number of dimensions used to construct a tensor must be equal to the
+    // rank of the tensor.
+    static_assert((sizeof...(otherDimensions) + 1 == dim_),
+                  "Number of dimensions given does not match rank");
+    dimensions_ = {firstDimension, otherDimensions...};
+
+    Index size = std::accumulate(dimensions_.begin(), dimensions_.end(), 1,
+                                 std::multiplies<Index>());
+    storage_.resize(size);
+    offsets_[0] = 1;
+    std::partial_sum(dimensions_.begin(), dimensions_.end() - 1,
+                     offsets_.begin() + 1, std::multiplies<Index>());
```

Comment:
Suddenly you became a functional programmer ;)

Reply:
yes,  I am starting to like this functional style. :)